test/e2e/cloud-provider-azure.go

```diff
@@ -0,0 +1,171 @@
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest/to"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	helmAction "helm.sh/helm/v3/pkg/action"
+	helmLoader "helm.sh/helm/v3/pkg/chart/loader"
+	helmCli "helm.sh/helm/v3/pkg/cli"
+	helmVals "helm.sh/helm/v3/pkg/cli/values"
+	helmGetter "helm.sh/helm/v3/pkg/getter"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// InstallCloudProviderAzureHelmChart installs the official cloud-provider-azure helm chart
+func InstallCloudProviderAzureHelmChart(ctx context.Context, input clusterctl.ApplyClusterTemplateAndWaitInput, result *clusterctl.ApplyClusterTemplateAndWaitResult) {
+	By("Waiting for workload cluster kubeconfig secret")
+	Eventually(func() error {
+		client := input.ClusterProxy.GetClient()
+		secret := &corev1.Secret{}
+		key := crclient.ObjectKey{
+			Name:      fmt.Sprintf("%s-kubeconfig", input.ConfigCluster.ClusterName),
+			Namespace: input.ConfigCluster.Namespace,
+		}
+		return client.Get(ctx, key, secret)
+	}, 20*time.Minute, 5*time.Second).Should(Succeed())
+	clusterProxy := input.ClusterProxy.GetWorkloadCluster(ctx, input.ConfigCluster.Namespace, input.ConfigCluster.ClusterName)
+	By("Waiting for nodes to come online indicating that the cluster is ready to accept work")
+	Eventually(func() error {
+		clientSet := clusterProxy.GetClientSet()
+		var runningNodes int
+		list, err := clientSet.CoreV1().Nodes().List(ctx, v1.ListOptions{})
+		if err != nil {
+			return err
+		}
+		for _, n := range list.Items {
+			if n.Status.Phase == corev1.NodeRunning {
+				runningNodes++
+			}
+		}
+		if runningNodes > 0 {
+			return nil
+		}
+		return err
+	}, 20*time.Minute, 5*time.Second).Should(Succeed())
+	kubeConfigPath := clusterProxy.GetKubeconfigPath()
+	os.Setenv("KUBECONFIG", kubeConfigPath)
+	By("Installing the correct version of cloud-provider-azure components via helm")
+	Eventually(func() error {
+		settings := helmCli.New()
+		actionConfig := new(helmAction.Configuration)
+		if err := actionConfig.Init(settings.RESTClientGetter(), "default", os.Getenv("HELM_DRIVER"), Logf); err != nil {
+			return err
+		}
+		i := helmAction.NewInstall(actionConfig)
+		i.RepoURL = "https://raw.githubusercontent.com/kubernetes-sigs/cloud-provider-azure/master/helm/repo"
+		i.ReleaseName = "cloud-provider-azure-oot"
+		cp, err := i.ChartPathOptions.LocateChart("cloud-provider-azure", helmCli.New())
+		if err != nil {
+			return err
+		}
+		p := helmGetter.All(settings)
+		valueOpts := &helmVals.Options{}
+		valueOpts.Values = []string{fmt.Sprintf("infra.clusterName=%s", input.ClusterProxy.GetName())}
+		if imageRegistryFromEnv := os.Getenv("IMAGE_REGISTRY"); imageRegistryFromEnv != "" {
```

Comment:
@CecileRobertMichon @jsturtevant the below is a proposed solution to ensuring these changes are back-compat with upstream cloud-provider-azure test-infra jobs that rely upon capz to test out-of-tree. This is how the current tests currently work:

1. `TEST_CCM=true` is set when calling `scripts/ci-entrypoint.sh`:
  - https://github.com/kubernetes/test-infra/blob/f74bc266fbbff04f30769fbfcd736634c90d3b5f/config/jobs/kubernetes-sigs/cloud-provider-azure/cloud-provider-azure-config.yaml#L185
2. capz maintains a script to set the CI-built image data here:
  - https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/1352fa2be39e1431f6beb62ecaa014471435b5ca/scripts/ci-entrypoint.sh#L91
  - https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/1352fa2be39e1431f6beb62ecaa014471435b5ca/scripts/ci-build-azure-ccm.sh#L43
  - https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/1352fa2be39e1431f6beb62ecaa014471435b5ca/scripts/ci-build-azure-ccm.sh#L34-L36
  - https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/1352fa2be39e1431f6beb62ecaa014471435b5ca/scripts/ci-build-azure-ccm.sh#L44

These env vars aren't as specifically named as we'd ideally want, but re-using them as-is allows us to reduce the change surface area (don't have to update capz CI scripts and/or test-infra job defs).

Reply:
the out of tree cloud provider PR tests and conformance tests don't run CAPZ e2e so they won't call `InstallCloudProviderAzureHelmChart` as far as I know. We may need to 1) modify ci-entrypoint.sh directly to do the helm install for out of tree cases, and 2) change conformance_test.go to use this custom control plane waiter too when testing out of tree

Reply:
/cc @lzhecheng

Reply:
I don't think we have to worry about the conformance stuff yet (in this PR), it seems the only template flavors that it currently uses are default and windows (not the external templates).

When we eventually get to the part of this effort where we're testing out-of-tree by default, then yeah, the conformance `ControlPlaneWaiters` will need to get updated as well.